generic `size`: avoid method static parameters and abstract type assert

nsajko opened the pull request:
The generic method `size(::AbstractArray, ::Any)` currently has:

* method static parameters

* an abstract type assert (`::Integer`) in the method body

PR #59442 aims to make this generic method be used for `Array` and `Memory`, by deleting the more specific methods.

It is my understanding that method static parameters and abstract type asserts can cause performance issues, when they're not optimized out, which happens when the argument types are not concretely inferred.

So I'm putting up this PR, which eliminates the method static parameters and the `typeassert`, to prevent any possible regressions from PR #59442.

Another thing that is necessary for preserving good abstract inference is to convert the index argument to `Int` before using it. This is correct as there can't be more than `typemax(Int)` dimensions anyway (the `N` type parameter to `AbstractArray` is an `Int` value).

nsajko committed:
generic `size`: avoid method static parameters and abstract type assert

The generic method `size(::AbstractArray, ::Any)` currently has:

* method static parameters

* an abstract type assert (`::Integer`) in the method body

PR #59442 aims to make this generic method be used for `Array` and
`Memory`, by deleting the more specific methods.

It is my understanding that method static parameters and abstract type
asserts can cause performance issues, when they're not optimized out,
which happens when the argument types are not concretely inferred.

So I'm putting up this PR, which eliminates the method static
parameters and the `typeassert`, to prevent any possible regressions
from PR #59442.

Another thing that is necessary for preserving good abstract inference
is to convert the index to `Int` before using it. This is correct
it there can't be more than `typemax(Int)` dimensions anyway.

nsajko commented:
Indeed, this PR is a prerequisite for PR #59442. Otherwise three of these effect inference tests fail:

https://github.com/JuliaLang/julia/blob/cdd4ac559647d89646f39acaed718945b730199d/Compiler/test/effects.jl#L792-L797

vtjnash reviewed (changes requested):
I don't like this. It changes clear code into a mess. Changing to simplified code like this would be okay though:
```julia
function size(t::AbstractArray, dim)
    d = Int(dim)::Int
    s = size(t)
    d <= length(s) ? s[d] : 1
end

```

nsajko committed:
simplify according to suggestion by vtjnash

Now this PR changes behavior: non-`Integer` values are allowed for the
second argument as long as they are convertible to `Int`.

nsajko committed:
typo

Co-authored-by: Jeff Bezanson <jeff.bezanson@gmail.com>

nsajko committed:
Merge branch 'master' into size_AbstractArray
